src/pydvl/influence/torch/functional.py

```diff
@@ -1093,3 +896,111 @@ def mat_mat_prod(x: torch.Tensor):
         shift_func=shift_func,
         mat_vec_device=operator.device,
     )
+
+
+def operator_spectral_approximation(
+    operator: "TensorOperator",
+    rank: int = 10,
+    krylov_dimension: Optional[int] = None,
+    tol: float = 1e-6,
+    max_iter: Optional[int] = None,
+    eigen_computation_on_gpu: bool = False,
+):
+    r"""
+    Calculates a low-rank approximation of an operator $H$ using the implicitly
+    restarted Lanczos algorithm, i.e.:
+
+    \[ H_{\text{approx}} = V D V^T\]
+
+    where \(D\) is a diagonal matrix with the top (in absolute value) `rank`
+    eigenvalues of the Hessian and \(V\) contains the corresponding eigenvectors.
+
+    Args:
+        operator: The operator to approximate.
+        rank: The number of eigenvalues and corresponding eigenvectors
+            to compute. Represents the desired rank of the Hessian approximation.
+        krylov_dimension: The number of Krylov vectors to use for the Lanczos
+            method. If not provided, it defaults to
+            \( \min(\text{model.n_parameters},
+                \max(2 \times \text{rank_estimate} + 1, 20)) \).
+        tol: The stopping criteria for the Lanczos algorithm, which stops when
+            the difference in the approximated eigenvalue is less than `tol`.
+            Defaults to 1e-6.
+        max_iter: The maximum number of iterations for the Lanczos method. If
+            not provided, it defaults to \( 10 \cdot \text{model.n_parameters}\).
+        eigen_computation_on_gpu: If True, tries to execute the eigen pair
+            approximation on the provided device via [cupy](https://cupy.dev/)
+            implementation. Ensure that either your model is small enough, or you
+            use a small rank_estimate to fit your device's memory. If False, the
+            eigen pair approximation is executed on the CPU with scipy's wrapper to
+            ARPACK.
+
+    Returns:
+        [LowRankProductRepresentation]
+            [pydvl.influence.torch.functional.LowRankProductRepresentation]
+            instance that contains the top (up until rank_estimate) eigenvalues
+            and corresponding eigenvectors of the Hessian.
+    """
+
+    if operator.input_size == 1:
+        # in the trivial case, return early
+        eigen_vec = torch.ones((1, 1), dtype=operator.dtype, device=operator.device)
+        eigen_val = operator.apply(eigen_vec).squeeze()
+        return LowRankProductRepresentation(eigen_val, eigen_vec)
+
+    torch_dtype = operator.dtype
+
+    if eigen_computation_on_gpu:
+        try:
+            import cupy as cp
+            from cupyx.scipy.sparse.linalg import LinearOperator, eigsh
+            from torch.utils.dlpack import from_dlpack, to_dlpack
+        except ImportError as e:
+            raise ImportError(
+                f"Try to install missing dependencies or set eigen_computation_on_gpu "
```

Comment:
It's just cupy that might be missing because torch is a dependency of pydvl, right? Then maybe we could have a more specific error message that says which package to install or even shows the pip command.